### ISSUE TITLE
1 ms ping accuracy

### ISSUE THREAD

**Issue by neszt** [tonydm/smokeping-OpenSSHMikrotikRouterOSPing#4]
The graphs show that the Mikrotik rounds the ping times and so on the one hand it is very ugly and on the other hand not accurate enough that way. Couldn't somehow get the Mikrotik to ping more accurately?

I found in Mikrotik 6.44 changelog:
```
speedtest - added "/tool speed-test" for ping latency, jitter, loss and TCP and UDP download, upload speed measurements (CLI only);
```

This gives the following output:
```
[admin@mikrotik] > /tool speed-test 192.168.88.2 duration=5 without-paging
          status: ping
  time-remaining: 55s
            loss: 100% (1/1)

              status: ping
      time-remaining: 54s
    ping-min-avg-max: 860us / 1.09ms / 3.93ms
  jitter-min-avg-max: 0s / 352us / 2.97ms
                loss: 0% (0/21)

              status: ping
      time-remaining: 53s
    ping-min-avg-max: 856us / 1.04ms / 3.93ms
  jitter-min-avg-max: 0s / 257us / 2.97ms
                loss: 0% (0/41)

              status: ping
      time-remaining: 52s
    ping-min-avg-max: 852us / 1.01ms / 3.93ms
  jitter-min-avg-max: 0s / 191us / 2.97ms
                loss: 0% (0/61)

              status: ping
      time-remaining: 51s
    ping-min-avg-max: 852us / 989us / 3.93ms
  jitter-min-avg-max: 0s / 159us / 2.97ms
                loss: 1.23% (1/81)
```

**Comment by tonydm** [comment]
@neszt 

Thank you for your interest in my Smokeping probe contribution.  Mikrotik rounding the ping result isn't super precise, but it's also not uncommon among ping utilities (see examples below).  To have microsecond (us) or 1 or more millisecond precision decimal results from a Mikrotik ping you would have to take that up with Mikrotik and it's software engineers.  My probe has no direct association with Mikrotik RouterOS, it's simply a tool/probe written to integrate with Smokeping which uses the Mikrotik ping tool.

Some would argue that having microseconds (us or ms.us) results isn't absolutely a deal breaker or even necessary to get a good understanding of a links latency or packet loss.  One of the nice things about Smokeping is the "smoke" that can tell you at a glace what the Std Deviation looks like.  There are many other tools that can give you more precision.  One way ping results can be had using  PerfSonar's OWAMP (One Way Ping).  Search google and you'll get links to their github page as well as their website.

https://www.perfsonar.net/
https://software.internet2.edu/owamp/owping.man.html
https://github.com/perfsonar/owamp
https://github.com/perfsonar

You can use the FPing probe included with Smokeping to get decimal point precision.  But, of course, that sources the pings from the server that is running the Smokeping daemon and negates the benefit of sourcing the pings from the Edgerouter's Public IP, a VLAN gateway, etc and introduces additional "noise" in the ping results.  However, in some cases that's exactly what you want, the routed result.

As far as the `/tool speed-test` you reference.  It is interesting.  However, it is meant to be used against another MIkrotik's BTest server.

#### Taken from Mikrotik's WiKi Page:  https://wiki.mikrotik.com/wiki/Manual:Tools/Speed_Test
```
The Speed Test is an easy test tool for measuring ping, jitter, TCP and UDP throughput from one MikroTik device, to another.
"speed-test" command is based on Ping Tool and Bandwidth Test. In order to use this command - Bandwidth test server needs to be accessible.
```

Having said that, using it against a generic endpoint, 1.1.1.1 for example, does give results (see below).  I played around with creating a test probe.  It kind of works.  The problem is that you can't specify the ping count, packet size, interface or source IP.  These are config params valuable to a ping's result.  Also, as seen from the `/tool speed-test` result below, when pinging another IP on the same router it can result in a microsecond result.  Smokeping does not process these as microseconds (us) but as milliseconds (ms).  This is handled by the Smokeping.pm source file.  You would have to direct your question to Tobias (the author of Smokeping) to make any changes to the source or fork the project and make the needed changes yourself.

Since this is an issue not with this probe, but with Mikrotik's Ping tool, this issue will be closed.

Very best :)

Tony

**# Microsoft Windows 10 cmd shell**
```
Microsoft Windows [Version 10.0.18363.1139]
(c) 2019 Microsoft Corporation. All rights reserved.

C:\Users\Tony>ping 1.1.1.1

Pinging 1.1.1.1 with 32 bytes of data:
Reply from 1.1.1.1: bytes=32 time=143ms TTL=53
Reply from 1.1.1.1: bytes=32 time=107ms TTL=53
Reply from 1.1.1.1: bytes=32 time=55ms TTL=53
Reply from 1.1.1.1: bytes=32 time=104ms TTL=53

Ping statistics for 1.1.1.1:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 55ms, Maximum = 143ms, Average = 102ms

C:\Users\Tony>
```

**# Linux shell ping using fping**
```
tonyd@deepthot:~$ fping -s 1.1.1.1 -c 20
1.1.1.1 : [0], 84 bytes, 117 ms (117 avg, 0% loss)
1.1.1.1 : [1], 84 bytes, 110 ms (113 avg, 0% loss)
1.1.1.1 : [2], 84 bytes, 134 ms (120 avg, 0% loss)
1.1.1.1 : [3], 84 bytes, 148 ms (127 avg, 0% loss)
1.1.1.1 : [4], 84 bytes, 122 ms (126 avg, 0% loss)
1.1.1.1 : [5], 84 bytes, 145 ms (129 avg, 0% loss)
1.1.1.1 : [6], 84 bytes, 137 ms (130 avg, 0% loss)
1.1.1.1 : [7], 84 bytes, 140 ms (131 avg, 0% loss)
1.1.1.1 : [8], 84 bytes, 37.0 ms (121 avg, 0% loss)
1.1.1.1 : [9], 84 bytes, 103 ms (119 avg, 0% loss)
1.1.1.1 : [10], 84 bytes, 105 ms (118 avg, 0% loss)
1.1.1.1 : [11], 84 bytes, 111 ms (117 avg, 0% loss)
1.1.1.1 : [12], 84 bytes, 121 ms (117 avg, 0% loss)
1.1.1.1 : [13], 84 bytes, 124 ms (118 avg, 0% loss)
1.1.1.1 : [14], 84 bytes, 104 ms (117 avg, 0% loss)
1.1.1.1 : [15], 84 bytes, 104 ms (116 avg, 0% loss)
1.1.1.1 : [16], 84 bytes, 99.0 ms (115 avg, 0% loss)
1.1.1.1 : [17], 84 bytes, 104 ms (115 avg, 0% loss)
1.1.1.1 : [18], 84 bytes, 61.7 ms (112 avg, 0% loss)
1.1.1.1 : [19], 84 bytes, 71.2 ms (110 avg, 0% loss)

1.1.1.1 : xmt/rcv/%loss = 20/20/0%, min/avg/max = 37.0/110/148

       1 targets
       1 alive
       0 unreachable
       0 unknown addresses

       0 timeouts (waiting for response)
      20 ICMP Echos sent
      20 ICMP Echo Replies received
       0 other ICMP received

 37.0 ms (min round trip time)
 110 ms (avg round trip time)
 148 ms (max round trip time)
       19.088 sec (elapsed real time)
```

**# Mikrotik Ping**
```
[tonyd@MikroTik] > ping 1.1.1.1 count=20 size=56
  SEQ HOST                                     SIZE TTL TIME  STATUS                                                      
    0 1.1.1.1                                    56  54 116ms
    1 1.1.1.1                                    56  54 133ms
    2 1.1.1.1                                    56  54 107ms
    3 1.1.1.1                                    56  54 123ms
    4 1.1.1.1                                    56  54 119ms
    5 1.1.1.1                                    56  54 134ms
    6 1.1.1.1                                    56  54 145ms
    7 1.1.1.1                                    56  54 121ms
    8 1.1.1.1                                    56  54 131ms
    9 1.1.1.1                                    56  54 113ms
   10 1.1.1.1                                    56  54 122ms
   11 1.1.1.1                                    56  54 123ms
   12 1.1.1.1                                    56  54 106ms
   13 1.1.1.1                                    56  54 136ms
   14 1.1.1.1                                    56  54 122ms
   15 1.1.1.1                                    56  54 145ms
   16 1.1.1.1                                    56  54 133ms
   17 1.1.1.1                                    56  54 99ms 
   18 1.1.1.1                                    56  54 120ms
   19 1.1.1.1                                    56  54 137ms
    sent=20 received=20 packet-loss=0% min-rtt=99ms avg-rtt=124ms max-rtt=145ms 

[tonyd@MikroTik] >
```

**# Mikrotik /tool speed-test**
```
[tonyd@MikroTik] > /tool speed-test address=<my_remote_mikrotik_router> user=<user> password=<password>
      status: done
      time-remaining: 0s
    ping-min-avg-max: 15ms / 114ms / 142ms
  jitter-min-avg-max: 107us / 7.23ms / 65ms
                loss: 0% (0/200)
        tcp-download: 197.4Mbps local-cpu-load:36%
          tcp-upload: 23.79Mbps local-cpu-load:20% remote-cpu-load:0%
          udp-upload: 27.10Mbps local-cpu-load:7% remote-cpu-load:0%

[tonyd@MikroTik] > /tool speed-test address=1.1.1.1                                    
              status: done
      time-remaining: 0s
    ping-min-avg-max: 21.6ms / 54.0ms / 134ms
  jitter-min-avg-max: 24us / 10.9ms / 103ms
                loss: 0% (0/200)

```

```
[tonyd@MikroTik] > /tool speed-test address=10.20.0.1                                  
              status: done
      time-remaining: 0s
    ping-min-avg-max: 228us / 247us / 420us
  jitter-min-avg-max: 0s / 10us / 180us
                loss: 0% (0/200)
        tcp-download: authentication failed
          tcp-upload: authentication failed
        udp-download: authentication failed
          udp-upload: authentication failed

[tonyd@MikroTik] >
```
